File: frontend/src/types/migration.ts
import type { Transformation } from "./transformation";
import type { PipelineStep, ColumnDef } from "./pipeline";

export type MigrationStatus =
  | "draft"
  | "testing"
  | "running"
  | "completed"
  | "failed"
  | "cancelled";

export interface Migration {
  id: string;
  name: string;
  source_connection_id: string;
  target_connection_id: string;
  source_table: string;
  target_table: string;
  source_query: string;
  source_schema: ColumnDef[];
  status: MigrationStatus;
  truncate_target: boolean;
  rows_processed: number | null;
  total_rows: number | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
  transformations: Transformation[];
  pipeline_steps: PipelineStep[];
  // Frontend-only display fields (not persisted)
  source_label?: string;
  target_label?: string;
  source_description?: string;
  target_description?: string;
}

export interface MigrationCreate {
  name: string;
  source_connection_id: string;
  target_connection_id: string;
  source_table: string;
  target_table: string;
  source_query?: string;
  source_schema?: ColumnDef[];
  truncate_target?: boolean;
}

export interface MigrationTestResult {
  before: Record<string, unknown>[];
  after: Record<string, unknown>[];
}

File: backend/src/vonnegut/models/migration.py
from typing import Literal

from pydantic import BaseModel, field_validator

from vonnegut.models.pipeline import PipelineStepResponse
from vonnegut.models.transformation import TransformationResponse

MigrationStatusType = Literal[
    "draft", "testing", "running", "completed", "failed", "cancelled"
]


class MigrationCreate(BaseModel):
    name: str
    source_connection_id: str
    target_connection_id: str
    source_table: str
    target_table: str
    source_query: str = ""
    source_schema: list[dict] = []
    truncate_target: bool = False

    @field_validator("name", "source_table", "target_table", mode="before")
    @classmethod
    def trim_strings(cls, v: str) -> str:
        return v.strip() if isinstance(v, str) else v


class MigrationUpdate(BaseModel):
    name: str | None = None
    source_table: str | None = None
    target_table: str | None = None
    source_query: str | None = None
    source_schema: list[dict] | None = None
    truncate_target: bool | None = None

    @field_validator("name", "source_table", "target_table", mode="before")
    @classmethod
    def trim_strings(cls, v: str | None) -> str | None:
        return v.strip() if isinstance(v, str) else v


class MigrationResponse(BaseModel):
    id: str
    name: str
    source_connection_id: str
    target_connection_id: str
    source_table: str
    target_table: str
    source_query: str = ""
    source_schema: list[dict] = []
    status: MigrationStatusType
    truncate_target: bool
    rows_processed: int | None
    total_rows: int | None
    error_message: str | None
    created_at: str
    updated_at: str
    transformations: list[TransformationResponse] = []
    pipeline_steps: list[PipelineStepResponse] = []
